### PR TITLE
Support for various clients

### PR DESCRIPTION
**Pull request by Shivam010**
Moving forward on the proposed work in the issue https://github.com/nitishm/go-rejson/issues/27#issuecomment-453861407

The project now, is all set to implement the ReJSON commands for various clients.

Checklist to implement `gomodule/redigo` client:

- [x] Interface setup and other stuffs to be used during client implementation
- [x] ~~Abstracting options for other commands like `JSON.ARRINDEX` and `JSON.ARRPOP`, as done for`JSON.GET` and `JSON.SET`~~ 
> **Edited:** I think their abstraction is not much required, they are very well maintained rigthnow
- [x] Adding proper examples and details in comments so that the godoc of the repo will be improved
- [x] Implementation of commands in `gomodule/redigo`
- [x] Adding proper test cases for clients

Similarly, for other clients:

- [x] `gomodule/redigo`
- [x] `go-redis/redis` 

Support for `mediocregopher/radix` is moved to the future roadmap

### CODE REVIEW

**Comment by Shivam010** [comment]
I think that the library should consider use of underscore, since it clearly signifies the parent-child relationship concept. 

_Just a suggestion_

**Comment by nitishm** [comment]
> > golint forces not to use underscores in Go names
> 
> I think that we should **consider using underscore**, since it clearly signifies the parent-child relationship concept. i.e. should increase the min-confidence of golint to 1
> 
> **_Just a suggestion_**

If you can make golint pass by setting min-confidence 1, then go for it !

**Comment by nitishm** [comment]
@Shivam010 Is this still a Work In Progress or are you done with supporting redigo ?

File: clients/redigo.go [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+//
+func (r *Redigo) JSONGet(key, path string, opts ...rjs.GetOption) (res interface{}, err error) {
+
+	if len(opts) > 4 {
+		return nil, rjs.TooManyOptionals
+	}
+	args := make([]interface{}, 0)
+	args = append(args, key)
+
+	args = append(args, path)
```
Comment (nitishm) [inline clients/redigo.go]: You can combine the two appends,
```
args = append(args, key, path)
```

File: clients/redigo.go [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+//
+func (r *Redigo) JSONArrIndex(key, path string, jsonValue interface{}, optionalRange ...int) (
+	res interface{}, err error) {
+
+	args := []interface{}{key, path, jsonValue}
+
+	ln := len(optionalRange)
+	switch {
+	case ln > 2:
+		return nil, rjs.NeedAtleastOneArg
```
Comment (nitishm) [inline clients/redigo.go]: `rjs.NeedAtleastOneArg` doesn't suit the case for `ln > 2`

Comment (nitishm) [reply clients/redigo.go]: at most ??

File: clients/redigo.go [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+// ReJSON syntax:
+// 	JSON.ARRINDEX <key> <path> <json-scalar> [start [stop]]
+//
+func (r *Redigo) JSONArrIndex(key, path string, jsonValue interface{}, optionalRange ...int) (
+	res interface{}, err error) {
+
+	args := []interface{}{key, path, jsonValue}
+
+	ln := len(optionalRange)
+	switch {
```
Comment (nitishm) [inline clients/redigo.go]: Do we need this switch statement ? You could just do `args = append(args, optionalRange...)`.

File: clients/redigo.go [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+	if len(opts) > 1 {
+		return nil, rjs.TooManyOptionals
+	}
+	args := make([]interface{}, 0, 5)
+	args = append(args, key, path, obj)
+
+	if len(opts) == 1 {
+		args = append(args, opts[0].Value()...)
+	}
+	name, args, err := rjs.CommandBuilder(rjs.ReJSONCommand_SET, args...)
```
Comment (nitishm) [inline clients/redigo.go]: > rjs.ReJSONCommand_SET

Not sure how I feel about mixing underscores with Camel-case, but I suppose its ok for now.

File: clients/redigo.go [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+func (r *Redigo) JSONNumIncrBy(key, path string, number int) (res interface{}, err error) {
+
+	name, args, err := rjs.CommandBuilder(rjs.ReJSONCommand_NUMINCRBY, key, path, number)
+	if err != nil {
+		return nil, err
+	}
+	return r.Conn.Do(name, args...)
+}
+
+// JSONNumMultBy to increment a number by provided amount
```
Comment (nitishm) [inline clients/redigo.go]: Seems to be a copy paste from JSONNumIncrBy. Might want to reword this to "... multiply a number ..."

File: clients/redigo.go [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+// 	JSON.ARRAPPEND <key> <path> <json> [json ...]
+//
+func (r *Redigo) JSONArrAppend(key, path string, values ...interface{}) (res interface{}, err error) {
+
+	if len(values) == 0 {
+		return nil, rjs.NeedAtleastOneArg
+	}
+	args := make([]interface{}, 0)
+	args = append(args, key, path)
+	args = append(args, values...)
```
Comment (nitishm) [inline clients/redigo.go]: Combine the two appends.

File: clients/redigo.go [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+// 	JSON.ARRINSERT <key> <path> <index> <json> [json ...]
+//
+func (r *Redigo) JSONArrInsert(key, path string, index int, values ...interface{}) (res interface{}, err error) {
+
+	if len(values) == 0 {
+		return nil, rjs.NeedAtleastOneArg
+	}
+	args := make([]interface{}, 0)
+	args = append(args, key, path, index)
+	args = append(args, values...)
```
Comment (nitishm) [inline clients/redigo.go]: Combine the appends

File: rjs/constants.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	ReJSONCommand_GET:       commandJSONGet,
+	ReJSONCommand_DEL:       commandJSONgeneric,
+	ReJSONCommand_MGET:      commandJSONMGet,
+	ReJSONCommand_TYPE:      commandJSONgeneric,
+	ReJSONCommand_NUMINCRBY: commandJSONNumIncrBy,
+	ReJSONCommand_NUMMULTBY: commandJSONNumMultBy,
+	ReJSONCommand_STRAPPEND: commandJSONStrAppend,
+	ReJSONCommand_STRLEN:    commandJSONgeneric,
+	ReJSONCommand_ARRAPPEND: commandJSONArrAppend,
+	ReJSONCommand_ARRLEN:    commandJSONgeneric,
```
Comment (nitishm) [inline rjs/constants.go]: commandJSONgeneric -> commandJSONGeneric. Keeps it in camel-case

File: rjs/helper.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
+package rjs
+
+import "fmt"
+
+// BytesToString converts each byte in a byte slice into character, else panic out
+func BytesToString(lst interface{}) (str string) {
```
Comment (nitishm) [inline rjs/helper.go]: Is this used anywhere else other than `package rjs` ? If not you dont need to make it exportable, since it's just a helper

File: rjs/constants.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
+package rjs
+
+import "fmt"
+
+// Generic and Constant Errors to be returned to client to maintain stability
+var (
+	ErrInternal       = fmt.Errorf("error: internal client error")
+	NoClientSet       = fmt.Errorf("no redis client is set")
```
Comment (nitishm) [inline rjs/constants.go]: Prefixing this with `Err` might bring more clarity.

File: clients/redigo.go (hunk 3 above)
Comment (Shivam010) [reply clients/redigo.go]: Yes, because the function, ArrIndex can only accept at most 2 optional fields.

File: clients/redigo.go (hunk 1 above)
Comment (Shivam010) [reply clients/redigo.go]: Done

File: clients/redigo.go (hunk 5 above)
Comment (Shivam010) [reply clients/redigo.go]: Done

File: clients/redigo.go (hunk 6 above)
Comment (Shivam010) [reply clients/redigo.go]: Actually, we can't.
`append` can only use either a non-variadic or a variadic parameter, at a time.
So, combining them:
```go
 args = append(args, key, path, values...)
```
will pops error.

File: clients/redigo.go (hunk 7 above)
Comment (Shivam010) [reply clients/redigo.go]: https://github.com/nitishm/go-rejson/pull/29#discussion_r250390783

File: rjs/constants.go (hunk 8 above)
Comment (Shivam010) [reply rjs/constants.go]: Sure

File: rjs/helper.go (hunk 9 above)
Comment (Shivam010) [reply rjs/helper.go]: Actually, is used in `package clients` only.

Comment (Shivam010) [reply rjs/helper.go]: Hence, is required to be exported, if we keep it in `helper.go`

File: rjs/constants.go (hunk 10 above)
Comment (Shivam010) [reply rjs/constants.go]: Done

File: clients/redigo.go (hunk 4 above)
Comment (Shivam010) [reply clients/redigo.go]: It is not, general that the under_score gives more clarity over camelCase, it's just a perspective of thinking. 
We can remove it too, I have no problem.

**Comment by Shivam010** [comment]
@nitishm 
I think, now the project is ready for its deployment, with two redis client support, 
* `gomodule/redigo`
* `go-redis/redis`

The coding part for both the clients are done, only the examples are left to implement.

And also, the use of underscores is removed from the project 😅

File: clients/redigo.go [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+}
+
+// JSONSet used to set a json object
+//
+// ReJSON syntax:
+// 	JSON.SET <key> <path> <json>
+// 			 [NX | XX]
+//
+func (r *Redigo) JSONSet(key string, path string, obj interface{}, opts ...rjs.SetOption) (
+	res interface{}, err error) {
```
Comment (nitishm) [inline clients/redigo.go]: [Nit]: Was this a gofmt thing or did you purposely move this to the next line ?

File: clients/redigo.go [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+	name, args, err := rjs.CommandBuilder(rjs.ReJSONCommandOBJKEYS, key, path)
+	if err != nil {
+		return nil, err
+	}
+	res, err = r.Conn.Do(name, args...)
+	if err != nil {
+		return
+	}
+	// JSON.OBJKEYS returns slice of string as slice of uint8
+	slc := make([]string, 0, 10)
```
Comment (nitishm) [inline clients/redigo.go]: Is the 10 arbitrary ?

File: clients/redigo.go [hunk 13]
```diff
@@ -0,0 +1,387 @@
+package clients
+
+import (
+	"fmt"
+	"github.com/Shivam010/go-rejson/rjs"
```
Comment (nitishm) [inline clients/redigo.go]: This will need to be changed : s/Shivam010/niitshm/ before pushing

File: doc.go [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+Installation
+
+
+To install and use ReJSON module, one must have the pre-requisites installed and setup. Follow the following steps :
+	go get github.com/nitishm/go-rejson
+	cd $GOPATH/src/github.com/nitishm/go-rejson
+	./install-redis-rejson.sh
+
+
+Examples
```
Comment (nitishm) [inline doc.go]: Are you planning to add examples ? If not lets remove this for now.

File: rejson.go [hunk 15]
```diff
@@ -1,510 +1,310 @@
...
-const (
-	// PopArrLast gives index of the last element for JSONArrPop
-	PopArrLast = -1
-	// DebugMemorySubcommand provide the corresponding MEMORY sub commands for JSONDebug
-	DebugMemorySubcommand = "MEMORY"
-	// DebugHelpSubcommand provide the corresponding HELP sub commands for JSONDebug
-	DebugHelpSubcommand = "HELP"
-	// DebugHelpOutput is the output of command JSON.Debug HELP <obj> [path]
-	DebugHelpOutput = "MEMORY <key> [path] - reports memory usage\nHELP                - this message"
+	"github.com/Shivam010/go-rejson/rjs"
```
Comment (nitishm) [inline rejson.go]: This will need to be changed : s/Shivam010/niitshm/ before pushing

File: rejson_test.go [hunk 16]
```diff
@@ -2,31 +2,229 @@ package rejson
 
 import (
 	"encoding/json"
+	"github.com/Shivam010/go-rejson/rjs"
```
Comment (nitishm) [inline rejson_test.go]: This will need to be changed : s/Shivam010/niitshm/ before pushing

File: rejson_test.go [hunk 17]
```diff
@@ -2,31 +2,229 @@ package rejson
...
+	_, _, err := rjs.CommandBuilder(1234, nil)
+	if err == nil {
+		t.Errorf("TestUnsupportedCommand() returned nil error")
+		return
+	}
 }
 
-func TestJSONSet(t *testing.T) {
-	conn, err := redis.Dial("tcp", ":6379")
+func TestRedigo(t *testing.T) {
```
Comment (nitishm) [inline rejson_test.go]: I really like how you structured this ! 💯

File: rejson_test.go [hunk 18]
```diff
@@ -2,31 +2,229 @@ package rejson
...
+	defer func() {
+		if err := red.FlushAll().Err(); err != nil {
+			t.Fatalf("goredis - failed to flush: %v", err)
+		}
+		if err := red.Close(); err != nil {
+			t.Fatalf("goredis - failed to communicate to redis-server: %v", err)
+		}
+	}()
+
+	t.Run("TestJSONSet", func(t *testing.T) {
```
Comment (nitishm) [inline rejson_test.go]: Wish they had a way to augment the `t.Run()` function with `func(t *testing.T, <extra-args>)`. Could have prevented repeating the test cases for each client.

File: clients/redigo.go (hunk 11 above)
Comment (Shivam010) [reply clients/redigo.go]: It's actually because of the linter `lll` which limits the length of a line to 120, hence, we are forced to move it to next line - to pass it.

File: clients/redigo.go (hunk 12 above)
Comment (Shivam010) [reply clients/redigo.go]: Yes, it's just an arbitrary call.

File: clients/redigo.go (hunk 13 above)
Comment (Shivam010) [reply clients/redigo.go]: Yes-yes, I have that in my mind.
Each and every import paths will be rechecked and confirmed before the final push, don't worry about that.

File: doc.go (hunk 14 above)
Comment (Shivam010) [reply doc.go]: Yes, I am planning to add some client setup examples to make it clear for the user.

File: rejson.go (hunk 15 above)
Comment (Shivam010) [reply rejson.go]: Don't worry about that.

File: rejson_test.go (hunk 17 above)
Comment (Shivam010) [reply rejson_test.go]: Thanks 😅

File: rejson_test.go (hunk 18 above)
Comment (Shivam010) [reply rejson_test.go]: I think the last commit https://github.com/nitishm/go-rejson/pull/29/commits/1c0b597271f6678a81ceb3c086d705e208529292 has fulfilled your wish. 😛
